src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs

```diff
@@ -218,93 +221,233 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
+            // the JIT to see that the code is unreachable and eliminate it when the platform does not have hardware accelerated.
+            if (Avx2.IsSupported)
+            {
+                if (offset < length)
                 {
-                    length--;
+                    Debug.Assert(length - offset >= Vector128<ushort>.Count);
+                    if ((((nint)Unsafe.AsPointer(ref searchSpace) + (nint)offset) & (nint)(Vector256<ushort>.Count - 1)) != 0)
+                    {
+                        // Not currently aligned to Vector256 (is aligned to Vector128); this can cause a problem for searches
+                        // with no upper bound e.g. String.wcslen. Start with a check on Vector128 to align to Vector256, 
+                        // before moving to processing Vector256.
+                        // If the input searchSpan has been fixed or pinned, this ensures we do not fault across memory pages 
+                        // while searching for an end of string.
+                        // If the searchSpan has not been fixed or pinned the GC can relocate it during the execution of this 
+                        // method, so the alignment only acts as best endeavour. The GC cost is likely to dominate over
+                        // the misalignment that may occur after; to we default to giving the GC a free hand to relocate and 
+                        // its up to the caller whether they are operating over fixed data.
+                        Vector128<ushort> values = Vector128.Create((ushort)value);
+                        Vector128<ushort> search = LoadVector128(ref searchSpace, offset);
+
+                        // Same method as below
+                        int matches = Sse2.MoveMask(Sse2.CompareEqual(values, search).AsByte());
+                        if (matches == 0)
+                        {
+                            // Zero flags set so no matches
+                            offset += Vector128<ushort>.Count;
+                        }
+                        else
+                        {
+                            // Find bitflag offset of first match and add to current offset
+                            return (int)(offset + (BitOperations.TrailingZeroCount(matches) / sizeof(char)));
+                        }
+                    }
 
-                    if (pCh[0] == value)
-                        goto Found;
+                    lengthToExamine = GetCharVector256SpanLength(offset, length);
+                    if (lengthToExamine > 0)
+                    {
+                        Vector256<ushort> values = Vector256.Create((ushort)value);
+                        do
+                        {
+                            Debug.Assert(lengthToExamine >= Vector256<ushort>.Count);
+
+                            Vector256<ushort> search = LoadVector256(ref searchSpace, offset);
```

Comment:
Likewise, what prevents an AV here if the GC relocates the underlying data and we are doing an unbounded search?

Reply:
Nothing; no guarantees are offered. If its a Span(*ptr) then its fixed data; but if its unfixed managed data (e.g. array) why would you be creating an unbounded Span longer than the array length over it?